plugins/modules/ndo_service_device_cluster.py

```diff
@@ -0,0 +1,641 @@
+    mso_schema_cache = MSOSchemaCache()
+    mso_template_cache = MSOTemplateCache()
+    if module.params.get("interface_properties") is not None:
+        interface_properties = get_interfaces_payload(
+            mso, mso_template, module.params.get("interface_properties"), references, mso_schema_cache, mso_template_cache
+        )
+    object_description = "Service Device Cluster"
+    path = "/deviceTemplate/template/devices"
+
+    existing_devices = mso_template.template.get("deviceTemplate", {}).get("template", {}).get("devices") or []
+
+    if name or uuid:
+        match = mso_template.get_object_by_key_value_pairs(
+            object_description,
+            existing_devices,
+            [KVPair("uuid", uuid) if uuid else KVPair("name", name)],
+        )
+        if match:
+            device_path = "{0}/{1}".format(path, match.index)
+            mso_template.update_config_with_template_and_references(match.details)
+            mso.existing = mso.previous = copy.deepcopy(match.details)
+    else:
+        mso.existing = mso.previous = [mso_template.update_config_with_template_and_references(device) for device in existing_devices]
+
+    if state == "present":
+        mso_values = dict(
+            name=name,
+            description=description,
+            deviceLocation="onPremise",
+            deviceMode=device_mode,
+            deviceType=device_type,
+            connectivityMode="advanced" if len(interface_properties) >= 3 else connectivity_mode.get(len(interface_properties)),
+            interfaces=interface_properties,
+        )
+
+        if match:
+            append_update_ops_data(ops, match.details, device_path, mso_values)
+            mso.sanitize(match.details, collate=True)
+        else:
+            mso.sanitize(mso_values)
+            ops.append(dict(op="add", path="{0}/-".format(path), value=mso.sent))
+
+    elif state == "absent":
+        if match:
+            ops.append(dict(op="remove", path=device_path))
+
+    if not module.check_mode and ops:
+        response = mso.request(mso_template.template_path, method="PATCH", data=ops)
+        devices = response.get("deviceTemplate", {}).get("template", {}).get("devices") or []
+        match = mso_template.get_object_by_key_value_pairs(object_description, devices, [KVPair("uuid", uuid) if uuid else KVPair("name", name)])
+        if match:
+            mso_template.update_config_with_template_and_references(match.details)
+            mso.existing = match.details
+        else:
+            mso.existing = {}
+    elif module.check_mode and state != "query":
+        mso_template.update_config_with_template_and_references(mso.proposed)
+        mso.existing = mso.proposed if state == "present" else {}
+
+    mso.exit_json()
+
+
+def get_interfaces_payload(mso, mso_template, interfaces, references, mso_schema_cache, mso_template_cache):
+    def get_object_uuid_from_schema(interface_type, uuid, interface):
```

Comment:
should we use more specific naming for the functions, since they are specific to the service device cluster interface part?